Add liquidity

kathaypacific opened the pull request:
fixes #44

kathaypacific committed:
Various tweaks to the UI

kathaypacific committed:
tweak

kathaypacific committed:
add skeleton functionality for add liquidity

kathaypacific committed:
add your stake

cajubelt commented on src/pages/Compound/useLiquidityRegistry.ts:
```diff
@@ -0,0 +1,79 @@
+      const token0Contract = new kit.web3.eth.Contract(ERC20_ABI as AbiItem[], token0Address)
+      const token0Symbol = await token0Contract.methods.symbol().call()
+
+      // TODO: is token1 always the RFP token? assume it is for now
```
it is not always!!!!

cajubelt replied:
token0 is just the first token when they are sorted alphabetically

kathaypacific replied:
oh yeah haha this was my quick hack to get it working. i'll use the list of RFP addresses we have in the compoundBotRegistry (or soon to be, farmBotRegistry) to find the correct token

cajubelt commented on src/pages/Compound/useLiquidityRegistry.ts:
```diff
@@ -0,0 +1,79 @@
+      liquidityPoolSummaries.push({
+        token0: new Token(ChainId.MAINNET, token0Address, 18, token0Symbol),
+        token1: new Token(ChainId.MAINNET, token1Address, 18, token1Symbol),
+        compoundBotSummary: {
```
can we replace "compoundBot" with "farmBot" in this repo to be more consistent with the other repos?

cajubelt replied:
I know I trash talked "farm bot" in the past. But it's used almost everywhere now in our code and docs

kathaypacific replied:
oh sure! i can make that change :)

kathaypacific committed:
refactor pool card to be reusable

kathaypacific committed:
make button font weight consistent

kathaypacific committed:
tidy up UI

kathaypacific committed:
add liquidity functionality

kathaypacific committed:
split into smaller components

kathaypacific committed:
adjust button styles

kathaypacific committed:
add header tabs

cajubelt commented on src/pages/App.tsx:
```diff
@@ -99,9 +101,10 @@ export default function App() {
-              </Route>
+              <Route exact strict path="/zap" component={Earn} />
+              <Route exact strict path="/pool" component={ProvideLiquidity} />
+              {/* TODO rename this */}
```
todo

kathaypacific replied:
this is done :)

cajubelt commented on src/pages/Swap/redirects.tsx:
```diff
@@ -7,7 +7,7 @@ import { ApplicationModal, setOpenModal } from '../../state/application/actions'
 
 // Redirects to swap but only replace the pathname
```
```suggestion
// Redirects to zap but only replace the pathname
```

kathaypacific replied:
also done! ended up rejigging some folders and making a file for this in the Zap folder

cajubelt commented on src/pages/ProvideLiquidity/AddLiquidityConfirm.tsx:
```diff
@@ -0,0 +1,180 @@
+    } catch (error: any) {
+      setAttemptingTxn(false)
+      // we only care if the error is something _other_ than the user rejected the tx
+      if (error?.code !== 4001) {
```
nit: magic numbers

cajubelt commented on src/pages/ProvideLiquidity/AddLiquidityForm.tsx:
```diff
@@ -0,0 +1,180 @@
+    {}
+  )
+
+  const atMaxAmounts: { [field in Field]?: TokenAmount } = [Field.CURRENCY_A, Field.CURRENCY_B].reduce(
+    (accumulator, field) => {
+      return {
+        ...accumulator,
+        [field]: maxAmounts[field]?.equalTo(parsedAmounts[field] ?? '0'),
+      }
+    },
+    {}
+  )
```
```suggestion
  const atMaxAmounts= (field: Field) => maxAmounts[field]?.equalTo(parsedAmounts[field] ?? '0')
```
code golf. didn't test. would need a slight refactor where it's used (replace atMaxAmounts[field] with atMaxAmounts(field)

cajubelt commented on src/pages/ProvideLiquidity/AddLiquidityForm.tsx:
```diff
@@ -0,0 +1,180 @@
+        onUserInput={onFieldAInput}
+        onMax={() => {
+          if (currencies[Field.CURRENCY_A]?.address === CELO[chainId as unknown as UbeswapChainId].address) {
+            onFieldAInput(Math.max(Number(maxAmounts[Field.CURRENCY_A]?.toExact() ?? '') - 0.01, 0).toString())
```
```suggestion
            onFieldAInput(Math.max(Number(maxAmounts[Field.CURRENCY_A]?.toExact() ?? '0') - 0.01, 0).toString())
```

cajubelt replied:
nit. not obvious what Number('') would evaluate to. (It evaluates to 0 so this is equivalent but clearer)

cajubelt replied:
also nit: magic numbers (took me some thinking to figure out what the 0.01 does)

kathaypacific replied:
honestly i took chunks from the existing code, and didn't take the time to understand each magic number as i wanted to get this done quickly. i probably should - but figure i have time later to do this (i didn't change details inside the chunks that i didn't dig into, so figure that this copy+paste is low risk) this is a good suggestion though, will apply :)

cajubelt commented on src/pages/ProvideLiquidity/AddLiquidityForm.tsx:
```diff
@@ -0,0 +1,180 @@
+      <StyledColumnCenter>
+        <Plus size="16" color={theme.text2} />
+      </StyledColumnCenter>
+      <CurrencyInputPanel
+        value={formattedAmounts[Field.CURRENCY_B]}
+        onUserInput={onFieldBInput}
+        onMax={() => {
+          if (currencies[Field.CURRENCY_B]?.address === CELO[chainId as unknown as UbeswapChainId].address) {
+            onFieldBInput(Math.max(Number(maxAmounts[Field.CURRENCY_B]?.toExact() ?? '') - 0.01, 0).toString())
+          } else {
+            onFieldBInput(maxAmounts[Field.CURRENCY_B]?.toExact() ?? '')
+          }
+        }}
+        showMaxButton={!atMaxAmounts[Field.CURRENCY_B]}
+        currency={token1}
+        id="add-liquidity-input-token-b"
+        disableCurrencySelect
+      />
```
nit: this could be more DRY

kathaypacific replied:
do you think so?? the 2 inputs look similar but most of the props are different (token0 vs token1) so i don't think abstracting it one more layer helps readability...

cajubelt replied:
yeah I think so. Reading this to look for bugs, I need to look at 9 different places to verify that the correct token was used (0/a or 1/b) and 4 different lines of logic to verify that it's the same in both places. If it were in a helper function, I'd only need to look at 3-4 parameters, and the logic would be guaranteed the same (since it's re-used). That's 9-10 fewer things to think about. 

But whatever, it's a nit.

cajubelt commented on src/pages/ProvideLiquidity/LiquidityCard.tsx:
```diff
@@ -0,0 +1,91 @@
+interface Props {
+  token0: Token
+  token1: Token
+  compoundBotSummary: CompoundBotSummaryBase
```
here's a place where "compoundBot" naming is used

kathaypacific replied:
i've gone through and renamed them just now

cajubelt reviewed (commented):
thank you!! 🙏 

I reviewed up until LiquidityCard.tsx and got tired 😴

lisaqyli commented:
I was playing with it. looks like remove liquidity is missing? Is it gonna be something included in this PR? If not, I'm happy to open a PR against your branch

kathaypacific commented:
@lisaqyli no this PR is just for adding liquidity. feel free to take the removing liquidity part :) i need to make some changes to this branch tonight, so maybe better to hold off branching off till then (or else you will have lots of conflicts!)

> I was playing with it. looks like remove liquidity is missing? Is it gonna be something included in this PR? If not, I'm happy to open a PR against your branch

kathaypacific committed:
rename compountBot to farmBot

kathaypacific committed:
rename file to farmBot

kathaypacific committed:
do not hardcode which order tokens come in from liquidity contract

kathaypacific committed:
show pool share correctly

kathaypacific committed:
move files around to make sense

kathaypacific committed:
add liquidity provider info block

kathaypacific committed:
Merge branch 'main' into kathy/add-liquidity

kathaypacific committed:
make onFieldInput more readable

cajubelt reviewed (approved):
nice job! 🥳 

A few thoughts from testing:
- would be cool if it switched to "manage" once the transaction goes through
- "your pool tokens" is less informative than the USD value of my liquidity
- there are still little question marks instead of token symbols
- we definitely need an info box explaining this feature (maybe one exists and I'm not seeing it because I've already provided liquidity?)
- mCUSD / mCUSD-mCEUR RFP LP is a real eyesore. Could we just use the mCUSD-mCEUR part? maybe "mCUSD-mCEUR zap pool"?

None of these observations are blocking for this PR, but most/all of them should probably be in GH issues and worked on soon-ish

lisaqyli commented on src/pages/App.tsx:
```diff
@@ -99,9 +101,9 @@ export default function App() {
-              <Route path="/">
-                <Earn />
-              </Route>
+              <Route exact strict path="/zap" component={Zap} />
```
nit: strict is bit unnecessary here? if someone types revo.market/zap/, the url would return 404

kathaypacific replied:
i have a default router so instead of a 404 you will be directed to the /zap page :) buuuut indeed `pool/` will be routed to `/zap`...this is actually what ubeswap has, i think it is safer than an inexact match in case we have a url like `/pools` in the future

lisaqyli commented:
I played with this branch and everything looks good.
I +1 on all of Charlie's thoughts, i had similar thoughts too. 
Also one more suggestion: i think the tab 'Provide' is bit vague as a title, how about naming it 'Pool' ?

kathaypacific committed:
ensure that token images show up in the liquidity card

kathaypacific committed:
update pool header title and liquidity card title

kathaypacific commented:
hey @cajubelt @lisaqyli thanks for your feedback!! to address your points above,
- i have updated the header item "provide" => "pool" (added this to the translations file to make it easier to change)
- Updated the pool name on the liquidity card to `{{token0Name}}-{{token1Name}} Zap Pool`
- Fixed the missing mcUSD token image on the liquidity card
- @cajubelt there is a info bit at the top of the pool screen if you disconnect your wallet, i actually used the same copy as what ubeswap has to explain pools. the link to the docs is just going to the generic Revo docs, lmk if there's anything revo specific that we should mention?

What we should do next?
- The manage view - i didn't implement a manage button here after liquidity is provided because we haven't built the remove liquidity yet (so manage = add liquidity, confusing)
- add a token image for the RFPs - should i just find a lightning logo? wasn't sure if it was important that we give it some thought. @lisaqyli maybe you have some thoughts?
- on the liquidity card i just have "your pool tokens" because this is what Ubeswap shows on their pool summary, but indeed it is nicer to have the usd value

lisaqyli commented:
I feel the logo of our RFP token is not a high priority task, the LP tokens on ubeswap (through custom import) also doesn't have a logo. my 2 cents, it might be a worthwhile task when we get several pools going since it might look confusing. Oh i have an idea, the lightning logo should be reserved for our governance token and all our LP tokens can be some sort of modification of the lightening logo.

kathaypacific committed:
add banner to warn beta
